tests/test_preprocess_pipeline.py

```diff
@@ -0,0 +1,115 @@
+import os.path
+from unittest import TestCase
+
+import numpy as np
+
+from aspire.estimation.noise import WhiteNoiseEstimator
+from aspire.source import ArrayImageSource
+from aspire.source.simulation import Simulation
+from aspire.utils.coor_trans import grid_2d, grid_3d
+from aspire.utils.filters import RadialCTFFilter, ScalarFilter
+from aspire.volume import Volume
+
+DATA_DIR = os.path.join(os.path.dirname(__file__), "saved_test_data")
+
+
+class PreprocessPLTestCase(TestCase):
+    def setUp(self):
+
+        self.L = 64
+        self.n = 128
+
+        self.noise_variance = 0.004
+        self.noise_filter = ScalarFilter(dim=2, value=self.noise_variance)
+
+        self.sim = Simulation(
+            L=self.L,
+            n=self.n,
+            unique_filters=[
+                RadialCTFFilter(defocus=d) for d in np.linspace(1.5e4, 2.5e4, 7)
+            ],
+            noise_filter=self.noise_filter,
+        )
+        self.imgs_org = self.sim.images(start=0, num=self.n)
+
+    def testPhaseFlip(self):
+        self.sim.phase_flip()
+        imgs_pf = self.sim.images(start=0, num=self.n)
+
+        # check energy conservation
+        self.assertTrue(np.allclose(self.imgs_org.norm(), imgs_pf.norm()))
+
+    def testDownsample(self):
+        # generate a 3D map with density decays as Gaussian function
+        g3d = grid_3d(self.L)
+        coords = np.array([g3d["x"].flatten(), g3d["y"].flatten(), g3d["z"].flatten()])
+        sigma = 0.2
+        vol = np.exp(-0.5 * np.sum(np.abs(coords / sigma) ** 2, axis=0))
+        vol = np.reshape(vol, g3d["x"].shape)
+        vols = Volume(vol)
+
+        # set noise to zero and CFT filters to unity for simulation object
+        noise_var = 0
+        noise_filter = ScalarFilter(dim=2, value=noise_var)
+        sim = Simulation(
+            L=self.L,
+            n=self.n,
+            vols=vols,
+            offsets=0.0,
+            amplitudes=1.0,
+            unique_filters=[
+                ScalarFilter(dim=2, value=1) for d in np.linspace(1.5e4, 2.5e4, 7)
+            ],
+            noise_filter=noise_filter,
+        )
+        # get images before downsample
+        imgs_org = sim.images(start=0, num=self.n)
+        # get images after downsample
+        max_resolution = 32
+        sim.downsample(max_resolution)
+        imgs_ds = sim.images(start=0, num=self.n)
+
```

Comment:
Also check that `imgs_ds[:, 16, 16] == imgs_org[:, 32, 32]` (perhaps up to some scaling factor).

Reply:
It looks that the downsampled value depends on the four neighboring points from the original images due to the linear interpolation algorithm in 2D grid. It is also not 1/4 of the sum of four points because the  2D grid is not shifted to the origin and the weights for four points are different. The scaling factor is not so obvious.

Reply:
So how is the grid aligned? Shouldn't half of the points in the original grid correspond to points on the downsampled grid?

Reply:
No. It is not. There is more complicated than this. See the http://fourier.eng.hmc.edu/e176/lectures/ch7/node7.html, it is obtained from the neighbouring grid points and weighted by the distances. If the distances are not the same value, it can not be recovered from the simple average. I tend to not add this test since the other indiect results are reasonable.

Reply:
Why is the grid not aligned? Seems like a reasonable thing to expect.

Reply:
For some reason, previously I checked  `imgs_ds[:, 15, 15]` == `imgs_org[:, 31, 31]`, thinking index starts from zero? 
I did it again today for four pairs:  `imgs_ds[:, 15, 15]`  vs `imgs_org [:, 30, 30]`, `imgs_ds[:, 15, 16]` vs `imgs_org[:, 30, 32]`,  `imgs_ds[:, 16, 15]` vs `imgs_org[:, 32, 30]`, and  `imgs_ds[:, 16, 16]` vs `imgs_org[:, 32, 32]`, all of them match each other.  For image 0, "0.22734155 0.2387177 0.2387177 0.2506631" are from the downsample points and  "0.22734158 0.2387177 0.2387177 0.2506631" for the origin points. Add the check in new commit 2f6e363.